src/Transaction/SignatureHash.php

```diff
@@ -31,47 +31,51 @@ public function __construct(TransactionInterface $transaction)
      * can be used.
      *
      * @param ScriptInterface $txOutScript
-     * @param $inputToSign
+     * @param int $inputToSignIdx
      * @param int $sighashType
      * @return Buffer
      * @throws \Exception
      */
-    public function calculate(ScriptInterface $txOutScript, $inputToSign, $sighashType = SignatureHashInterface::SIGHASH_ALL)
+    public function calculate(ScriptInterface $txOutScript, $inputToSignIdx, $sighashType = SignatureHashInterface::SIGHASH_ALL)
     {
-        $copy = $this->transaction->makeCopy();
+        $inputs = $this->transaction->getInputs();
+        $outputs = $this->transaction->getOutputs();
 
-        $inputs = $copy->getInputs();
-        $outputs = $copy->getOutputs();
-
-        if ($inputToSign > count($inputs)) {
+        if ($inputToSignIdx > count($inputs)) {
             throw new \Exception('Input does not exist');
         }
 
         // Default SIGHASH_ALL procedure: null all input scripts
         $inputCount = count($inputs);
         for ($i = 0; $i < $inputCount; $i++) {
-            $inputs->getInput($i)->setScript(new Script());
+            // null the script
+            $input = $inputs->getInput($i);
+            $inputs->setInput($i, new TransactionInput($input->getTransactionId(), $input->getVout(), new Script(), $input->getSequence()));
         }
 
-        $inputs->getInput($inputToSign)->setScript($txOutScript);
+        // set the $txOutScript
+        $inputToSign = $inputs->getInput($inputToSignIdx);
+        $inputs->setInput($inputToSignIdx, new TransactionInput($inputToSign->getTransactionId(), $inputToSign->getVout(), $txOutScript, $inputToSign->getSequence()));
         $math = Bitcoin::getMath();
 
         if ($math->bitwiseAnd($sighashType, 31) == SignatureHashInterface::SIGHASH_NONE) {
             // Set outputs to empty vector, and set sequence number of inputs to 0.
-            $copy->setOutputs(new TransactionOutputCollection());
+            $this->transaction->setOutputs(new MutableTransactionOutputCollection());
```

Comment:
I mean that for inputToSignIdx = 0, it might be sighash_none. the modifications are persisting past then. calculateHash should have no side effects. I'll have a look and see if the tests are meant to catch this..